ENH: disjoint set data structure

pmla opened the pull request:
<!-- 
Thanks for contributing a pull request! Please ensure that
your PR satisfies the checklist before submitting:
http://scipy.github.io/devdocs/dev/contributor/development_workflow.html#checklist-before-submitting-a-pr

Also, please name and describe your PR as you would write a
commit message:
https://docs.scipy.org/doc/numpy/dev/development_workflow.html#writing-the-commit-message

Note that we are a team of volunteers; we appreciate your
patience during the review process.

Again, thanks for contributing!
-->

#### What does this implement/fix?
<!--Please explain your changes.-->
This PR implements a [disjoint set](https://en.wikipedia.org/wiki/Disjoint-set_data_structure) data structure (also known as union-find). The `DisjointSet` class implements two methods only: `union` and `find`. Union performs "union by size" and implements path compression only (not path halving or splitting).
~I have written it in cython in order to get good performance, so that it can also be used by other SciPy components in future.~ Edit: it is now a pure python implementation.

I have put it in ~`sparse.csgraph`~ `cluster.hierarchy` based on @rgommers suggestion from the [dev list](https://mail.python.org/pipermail/scipy-dev/2020-July/024285.html):
```
Cluster, spatial and sparse all could make sense. If we foresee using this
internally in all those modules, I think we'd have to put it in `sparse` or
in `_lib`. The reason: no further dependencies between modules. cluster
already depends on spatial, and spatial depends on sparse.
```

Tests and documentation are included.

pmla commented:
Build failures appear to be unrelated.

sethtroisi commented on scipy/sparse/csgraph/_disjoint_set.pyx:
```diff
@@ -0,0 +1,130 @@
+        """Merge the subsets of `a` and `b`. The smaller subset (the child) is
+        merged into the the larger subset (the parent). If the subsets are of
+        equal size, the parent is determined by subset root with the smallest
+        index.
```
> If the subsets are of equal size, the parent is determined by subset root with the smallest index.

I don't see this logic implemented.

pmla replied:
This code puts the roots in sorted order:
```
        if b < a:
            a, b = b, a
```
Then, `b` can only become the parent if its size is larger than that of `a`:
```
if sizes[a] < sizes[b]:
```
This logic is also tested in  `test_equal_size_ordering`.

sethtroisi commented on scipy/sparse/csgraph/tests/test_disjoint_set.py:
```diff
@@ -0,0 +1,39 @@
+
+    parents = [dis.find(i) for i in range(n)]
+    assert_array_equal(np.arange(n), parents)
+
```
how about a more complicated join where find needs to recurse more than once, something like

```
def test_full_binary_tree():
    n = 16
    dis = DisjointSet(n)
    for i in range(0, 16, 2):
        assert not dis.union(i, i+1)
    for i in range(0, 16, 4):
        assert not dis.union(i+1, i+3)
    assert not dis.union(3, 7)

    parents = [dis.find(i) for i in range(n)]
    assert_array_equal([0,0,0,0,0,0,0,0,8,8,8,8,12,12,12,12], parents)
```

pmla replied:
Thanks for the suggestion. I have added a test to this effect.

sethtroisi commented on scipy/sparse/csgraph/_disjoint_set.pyx:
```diff
@@ -0,0 +1,130 @@
+    def __init__(DisjointSet self, np.intp_t n):
+        self.n = n
+        self.nc = n
+        self._sizes = np.ones(n, dtype=np.intp)
+        self._parents = np.arange(n, dtype=np.intp)
```
Maybe this is a v2 request, but the python disjointsets doesn't required you to know the number of nodes (it internally uses a dict).

This means you can use arbitrary indexes (e.g. coords in a n'd graph) and avoids the requirement of knowing up front how many items you'll want to add.

pmla replied:
Thanks for bringing this up. I did consider using a dict-based design as it offers more flexibility. I decided against it though because of the performance cost of using dicts rather than integer arrays.

The other argument for integer arrays is that functions in `sparse.csgraph` operate on fixed-size graphs with integer-indexed nodes, so I think the current design is more appropriate for the module.

sethtroisi commented on scipy/sparse/csgraph/tests/test_disjoint_set.py:
```diff
@@ -0,0 +1,53 @@
+
+    for k in 2**np.arange(kmax):
+        for i in range(0, n, 2 * k):
+            assert dis.union(i, i + k)
```
I love this test, let's do a couple more rounds, maybe `kmax = 10`?

If I can visualize correctly
`dis.union(i, i+k)` both `i` and `i+k` are current roots,
what about
`dis.union(i, i + k + (k-1))`

pmla replied:
I have bumped `kmax` to 10 as you suggested. I have also added randomized offsets, so we will get a mixture of root and non-root nodes.

sethtroisi commented on scipy/sparse/csgraph/tests/test_disjoint_set.py:
```diff
@@ -0,0 +1,53 @@
+    dis = DisjointSet(n)
+
+    for i in range(0, n, 2):
+        assert dis.union(i, i + 1)
```
it feels like you should be passing `(i + 1, i)` if you want to test a and b get swapped.

pmla replied:
Thanks. I have increased the size of the set and used randomized ordering of `i` and `i + 1` so that we get a mixture of input orderings.

sethtroisi commented on scipy/sparse/csgraph/_disjoint_set.pyx:
```diff
@@ -0,0 +1,130 @@
+
+
+cdef class DisjointSet:
+    """Disjoint set data structure [1]_.
```
This seems to render weird in the [docs](https://20588-1460385-gh.circle-artifacts.com/0/html-scipyorg/sparse.csgraph.html?highlight=csgraph#module-scipy.sparse.csgraph) maybe move to a 2nd line?

pmla replied:
I have moved the reference into the notes section. Hopefully that will fix it.

pmla commented:
> Anything else I should add to my checklist?

Checklist looks good. It is usually best to establish a broad consensus when new functionality is being added, so let's ask @tylerjereddy, @fuglede: you are `sparse` regulars with cython skills. Can I interest you in looking this PR over? Particular concerns of mine are:
* Suitability for inclusion in `sparse.csgraph`
* Have I done anything daft cython-wise.

pmla commented:
@sethtroisi btw, thanks for the great review!

ewmoore commented:
So right now this will expose the `IndexError` from numpy if you call `find` with an element equal to `n` or larger.  Is that the correct behavior, or should this produce some other error for this case?  

Let's say that it turns out that my initial `n` was too small, should there be away to resize the `DisjointSet` i have without redoing whatever other calculation I had to do the first time?  I suppose that this might not be possible at all.  It does look like resizing the two arrays and updating `n` and `nc` should do it though.

I guess a V2 behavior could be to dynamically resize the arrays you're using like a C++ vector will do.  I.e. starting with a default size and then increasing it using some heuristic to keep the unused size bounded while keeping the correct or nearly correct big-O  behavior by limiting the reallocations.   I suppose this could also be implemented via dicts or c++ maps too.

nmayorov commented on scipy/sparse/csgraph/_disjoint_set.pyx:
```diff
@@ -0,0 +1,137 @@
+
+    Parameters
+    ----------
+    n : int
```
Perhaps use terminology: `n -> n_nodes`, `nc -> n_components`.

Also, it should be mentioned somewhere that nodes indices are integers from `[0, 1, ..., n_nodes - 1]`.

pmla replied:
Thanks. I have made these changes.

nmayorov commented on scipy/sparse/csgraph/_disjoint_set.pyx:
```diff
@@ -0,0 +1,137 @@
+        parents[x] = parent
+        return parent
+
+    def union(DisjointSet self, np.intp_t a, np.intp_t b):
```
I believe `union` is always a noun, so it seems somewhat unconventional from Python code perspective.

How about considering `unite`, `join` or `merge` instead?

ewmoore replied:
This is `union` as in the union of sets.  Its the right name.

pmla replied:
@nmayorov I considered this. I also find `union` a little bit jarring but this is the most common in the literature. According to [Wikipedia](https://en.wikipedia.org/wiki/Disjoint-set_data_structure) another name for the disjoint set structure is a "merge-find set".

I will leave it as-is though.

nmayorov commented on scipy/sparse/csgraph/_disjoint_set.pyx:
```diff
@@ -0,0 +1,137 @@
+        return parent
+
+    def union(DisjointSet self, np.intp_t a, np.intp_t b):
+        """Merge the subsets of `a` and `b`. The smaller subset (the child) is
```
The first line should be limited to ``Merge the subsets of `a` and `b`.``

pmla replied:
Thanks. Fixed.

nmayorov commented on scipy/sparse/csgraph/_disjoint_set.pyx:
```diff
@@ -0,0 +1,137 @@
+
+        Parameters
+        ----------
+        a : int
```
Maybe simplify to 

```
a, b : int
    Node indices to merge.
```

pmla replied:
Thanks. I have made this simplification.

pmla commented:
> So right now this will expose the IndexError from numpy if you call find with an element equal to n or larger. Is that the correct behavior, or should this produce some other error for this case?

That was my original intention in the initial commit. In the current commit the requirement to specify `n` up front has been removed. It is now a pure python implementation using dicts rather than an cython array-based implementation. For this PR I propose that we stick to the python implementation in order to hammer out the interface.

> I suppose this could also be implemented via dicts or c++ maps too.

The class currently operates on integer node indices only. I think this is in line with the other functions in the `csgraph` module. It also means that it can be ported to C++ using maps instead of dicts in V2.

charris commented:
The python implementation linked in the original discussion of adding graph algorithms (2011) is in networkx is at https://github.com/networkx/networkx/blob/master/networkx/utils/union_find.py in case anyone is interested. Note that it merges by weight rather than height, both methods are efficient.

fuglede commented on scipy/sparse/csgraph/_disjoint_set.py:
```diff
@@ -0,0 +1,137 @@
+    def union(self, a, b):
+        """Merge the subsets of `a` and `b`.
+
+        The smaller subset (the child) is merged into the the larger subset
```
Typo: "the the"

pmla replied:
Thanks. Fixed.

fuglede commented on scipy/sparse/csgraph/_disjoint_set.py:
```diff
@@ -0,0 +1,137 @@
+
+    Initialize a disjoint set with 4 nodes:
+
+    >>> dis = DisjointSet(4)
```
The initializer signature seems to have changed along the way, so this one should result in an argument count mismatch.

pmla replied:
Good catch! Thanks.

fuglede commented:
API-wise, if I read @rgommers' comment on the mailing list correctly, he considered having it live in `sparse` directly, as opposed to `sparse.csgraph`; this might simplify the dependency graph. (Unless `sparse` and `sparse.csgraph` are considered equivalent from a dependency point of view(?))

fuglede commented:
Also API-wise, but more of a remark than a question/comment: While I'm not going to argue that this one is relevant in its own right (from the dev list discussion, it certainly seems like you and others have plenty of use cases already, and NetworkX shipping its own implementation is a good sign as well), but at the same time, it seems like this one is likely to be duplicated for performance (e.g. when inlining is desirable), sort of like fixed-size stacks and queues generally are. For instance, I guess you don't envision replacing the logic in the existing Kruskal-based MST solver, or the existing `connected_components`?

fuglede commented on scipy/sparse/csgraph/_disjoint_set.py:
```diff
@@ -0,0 +1,138 @@
+        merged : bool
+            `True` if `a` and `b` were in disjoint sets, `False` otherwise.
+        """
+        if not np.issubdtype(type(a), np.integer):
```
I take it you want to limit the API to integers, because it might be useful to restrict the inputs to correspond to indices of an array. Is that right? If so, two immediate questions:

* Right now `d = DisjointSet(); d.find(3)` produces an object whose `n_nodes` is 1; if this were ever changed to be array-based instead, could this be surprising behavior? That is, might one be led to expect that once a query is made for a given input `n`, then `m` is also an available node for all `0 < m < n`?

* Does it make sense that negative inputs are allowed?

fuglede replied:
But at the same time, if the intention is to potentially move to maps whose keys are fixed-size integers, there's potential API breakage here since you currently allow arbitrarily large input values. For complete future-proofing, it seems you would want to restrict them to belong to the appropriate range of values.

fuglede commented on scipy/sparse/csgraph/_disjoint_set.py:
```diff
@@ -0,0 +1,138 @@
+        parent = parents[x]
+        while parent != parents[parent]:
+            parent = parents[parent]
+        parents[x] = parent
```
You mention in the documentation that you implement path compression, but from a quick look, that doesn't quite seem to be the case (at least not in the sense I'm familiar with); while the parent is indeed updated for the input node `x`, it's not updated for anything else along the way (which would require a recursion, explicitly keeping track of the path, two iterations, or something along those lines).

pmla replied:
Another great catch. I have now implemented "path-halving" and updated the docs.

fuglede commented on scipy/sparse/csgraph/_disjoint_set.py:
```diff
@@ -0,0 +1,138 @@
+        if a == b:
+            return False
+
+        if b < a:
```
These two lines look redundant: The relative values of `a` and `b` never play a role; only their corresponding relative values in `self._sizes`.

fuglede replied:
But you could do the swap below and shave off a few lines; that is, have the entire thing boil down to:
```python
a = self.find(a)
b = self.find(b)
if a == b: return False

sizes = self._sizes
parents = self._parents
if sizes[a] < sizes[b]:
    a, b = b, a

parents[b] = a
sizes[a] += sizes[b]

self.n_components -= 1
return True
```

fuglede replied:
Ah, I now see that the first swap helps you break ties. (Is that important?)

pmla replied:
I want a tie-breaker in case `self._sizes[a] == self._sizes[b].
I have documented this in the docstring with the text

> If the subsets are of equal size, the parent is determined by subset root with the smallest index.

pmla replied:
Sorry, I should have reloaded this page before commenting.

I prefer it if ties are broken. It means that `union(a, b)` always produces the same behaviour as `union(b, a)`. I think I will leave it as-is. What do you think? When I do a C++ V2 this will be almost free.

fuglede replied:
Heh, looks like I just managed to squeeze in my realization the same minute as you wrote the explanation. But the logic sounds reasonable enough to me! It takes some flexibility away from the API, but I don't see a case where that's really an issue.

The phrasing of the last sentence in the documentation does seem ever so slightly off to me; could something like "If the subsets are of equal size, the parent is determined to be the subset whose root has the smallest index" be a bit clearer?

pmla replied:
I got rid of tie-breaking since I have changed the interface to allow arbitrary objects. Equality is difficult enough, so I don't want to start dealing with inequalities.

pmla commented:
Thanks for a great review @fuglede.

> API-wise, if I read @rgommers' comment on the mailing list correctly, he considered having it live in sparse directly, as opposed to sparse.csgraph; this might simplify the dependency graph. (Unless sparse and sparse.csgraph are considered equivalent from a dependency point of view(?))

I'm very happy to move it. Shall we wait for him to weigh in or just do it?

> ... it seems like this one is likely to be duplicated for performance (e.g. when inlining is desirable), sort of like fixed-size stacks and queues generally are. For instance, I guess you don't envision replacing the logic in the existing Kruskal-based MST solver, or the existing connected_components?

You are correct. Here I'm aiming for a publicly-accessible implementation which is also fast enough for some less demanding internal use cases. An example I had in mind is a PR of mine (#11502) where I use a disjoint set. That PR is already too big and the addition of the disjoint set is a distraction.

The challenge in this PR has been to find a balance between a flexible implementation ([as in networkx](https://github.com/networkx/networkx/blob/master/networkx/utils/union_find.py) ) where any hashable object can be used as a node, and a faster implementation where only integers can be used as keys.

fuglede commented:
> Shall we wait for him to weigh in or just do it?

I frankly can't claim to have enough of an overview to say what makes more sense.

Other than that, and the two minor inline comments on documentation and input type/range, LGTM!

charris commented:
> The challenge in this PR has been to find a balance between [fast and flexible]

I tussled with that too, back in the day, and also ended up with integers. They aren't the most transparent from the interface side, but fast and easy to code. My guess is that a more flexible interface could be made on top of the integers with a translation layer from objects to integers if needed. So maybe a fast cython/c++ base class/functions with a python interface.

charris commented:
I may have mentioned that I also like to keep the set elements in a circular list, the links are easy to manage when merging, and the entire set can be traversed starting from any element in it.

charris commented:
Here is a link to my old *.pyx implementation https://gist.github.com/charris/62f2dadc0ab597196635e8803eab786a

pmla commented:
After some thought I have revised the design so that any immutable object can be used as input. This was motivated by @fuglede's observation that this class will not be fast enough to replace inline code. In that light, restricting the input to integers feels like occupying a poor niche with neither optimal speed nor optimal flexibility.

In V2 I will rewrite in C++ backed by a python dict, and try to squeeze as much performance out as possible. I propose to save additional features for later PRs.

fuglede commented:
By "immutable", we really want "hashable", right? (Tuples are immutable, yet not necessarily hashable, and conversely, user defined types are hashable through `id`, yet not necessarily immutable.)

Just keeping C++ around for the traversal-related parts of the code sounds logical enough to me; and certainly a bit easier than having to wrestle hashes manually in calling code. For the C++ part, do you intend to use a hash map indexed by references to the Python objects, or something like that?

fuglede commented on scipy/sparse/csgraph/_disjoint_set.py:
```diff
@@ -0,0 +1,126 @@
+            # add node
+            self._sizes[x] = 1
+            self._parents[x] = x
+            self._indices[x] = len(self._indices)
```
Curious about your thoughts on this one: As far as I can tell, two objects will have different indices if and only if they are unequal. If so, you should be able to write things like `self._indices[a] == self._indices[b]` simply as `a == b`, but maybe there are other good reasons to keep `self._indices` around?

pmla replied:
The equality operator throws a ValueError if  e.g. `a` is a numpy int and `b` is a tuple with multiple elements.

Running them through the dict allows me to test for equality without this problem.

fuglede replied:
Ah, hm, what's a good example? `np.int(5) == (60, 80)` gives `False` as expected. But I see what you're saying, `__eq__` can be used for all sorts of magic; NumPy arrays can produce arrays as results, and things like ` np.ones((2, 3)) == np.ones((4, 5, 6))` will err in the future.

The bad examples I can come up with are all unhashable, though, so they can never end up in the data structure to begin with.

Except they can, if the user conjures their own type with a well-defined `__hash__` but poorly defined `__eq__`; supporting all such cases seems like it's hard in practice.

pmla replied:
Sorry, I should have specified that it needs to be e.g. a `np.int64`, rather than `np.int`.
If I change the type in your example I get bad behaviour:
```
>>> if np.int64(5) == (60, 80): print("hello")
... 
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
ValueError: The truth value of an array with more than one element is ambiguous. Use a.any() or a.all()
```

fuglede replied:
Really?

```python
Python 3.8.3 (default, Jul  2 2020, 16:21:59)
[GCC 7.3.0] :: Anaconda, Inc. on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import numpy as np
>>> np.int64(5) == (60, 80)
array([False, False])
>>> np.__version__
'1.19.1'
```
Tested on 1.18.1 and 1.18.5 as well.

fuglede replied:
So with that example, what I was hinting at above is that "`not in`" will fail for for `a = (60, 80)` and an `np.int64` if its hash happens to agree with `a`, even with `_indices` around:

```python
In [5]: d = DisjointSet()
In [6]: a = (60, 80)
In [7]: h = hash(a)  # Something like 1524157532169088664
In [9]: b = np.int64(h)  # Happens to satisfy hash(b) = h
In [10]: d.union(a, b)
ValueError
```

So, pretty edge-casey of course. I see how `_indices` allows you to go from not being able to work with cases where `__eq__` fails, to only having to worry about cases where `__eq__` fails but `__hash__` doesn't save the day.

This opens the door for some extremely subtle bugs, but they're edge-casey enough that it's probably not a problem at all, and the pros outweight the cons. Perhaps you even have a real life use case a la `np.int64` in mind yourself?

pmla replied:
Thanks for your comments here. I think we are in agreement now, so marking as resolved. Feel free to unresolve if not.

fuglede replied:
As a complete aside, I just now realized that my example actually doesn't blow up on Python 3.7 (but it does on Python 3.8) since, to my surprise, `hash(hash(a))` is not in general the same as `hash(a)` on Python 3.7. (Not sure if that's a CPython implementation detail, or what is going on though.)

pmla commented:
> By "immutable", we really want "hashable", right?

Absolutely - good catch. I have updated the docstrings.

nmayorov commented on scipy/sparse/csgraph/_disjoint_set.py:
```diff
@@ -0,0 +1,126 @@
+        Returns
+        -------
+        merged : bool
+            `True` if `a` and `b` were in disjoint sets, `False` otherwise.
```
I think ticks around True and False are unnecessary.

pmla replied:
Thanks. Fixed.

nmayorov commented on scipy/sparse/csgraph/_disjoint_set.py:
```diff
@@ -0,0 +1,126 @@
+
+    Notes
+    -----
+    This class implements the disjoint set [1]_, also known as the *union-find*
```
May change to "*union-find* or *merge-find*".

pmla replied:
I have changed the docs as proposed.

nmayorov commented on scipy/sparse/csgraph/_disjoint_set.py:
```diff
@@ -0,0 +1,126 @@
+    Notes
+    -----
+    This class implements the disjoint set [1]_, also known as the *union-find*
+    data structure. The *find* method implements the *path halving*
```
I guess "*find* method" should not be mentioned anymore. Perhaps "*find* operation implemented via `__getitem__`" or something like that.

pmla replied:
Good suggestion. I have made the change.

sethtroisi commented on scipy/sparse/csgraph/tests/test_disjoint_set.py:
```diff
@@ -0,0 +1,79 @@
+            yield tuple(node)
+
+
+def get_nodes(n, shuffle):
```
shuffle seems unused.

pmla replied:
Good catch. I have removed this.

pmla commented on scipy/sparse/csgraph/_disjoint_set.py:
```diff
@@ -0,0 +1,126 @@
+    Methods
+    -------
+    union
+    find
```
Thanks. Fixed.

pmla commented on scipy/sparse/csgraph/_disjoint_set.py:
```diff
@@ -0,0 +1,126 @@
+
+    Initialize a disjoint set:
+
+    >>> dis = DisjointSet()
```
I have made this change.

pmla commented on scipy/sparse/csgraph/_disjoint_set.py:
```diff
@@ -0,0 +1,126 @@
+            x = parents[x]
+        return x
+
+    def union(self, a, b):
```
I agree. I have renamed it. Hopcroft uses this terminology too.

nmayorov commented on scipy/sparse/csgraph/_disjoint_set.py:
```diff
@@ -0,0 +1,127 @@
+
+    Find root nodes:
+
+    >>> disjoint_set[2]
+    1
+    >>> disjoint_set['y']
+    3
```
Looks like a good idea to me.

pmla replied:
I have added a method called `connected` and integrated it into the unit tests.

pmla commented on scipy/sparse/csgraph/_disjoint_set.py:
```diff
@@ -0,0 +1,127 @@
+        self._parents = {}
+        self._indices = {}
+
+    def __getitem__(self, x):
```
I have renamed `a` and `b` to `x` and `y` instead.

pmla commented:
I have added an `__iter__` method and changed the `self._indices` from a `dict` to an `OrderedDict`.

Before merging I want to add tie-breaking back in to the `merge` method. I plan to use order of insertion to break ties. I will get on this later.

sethtroisi commented on scipy/sparse/csgraph/_disjoint_set.py:
```diff
@@ -0,0 +1,177 @@
+        self.n_components -= 1
+        return True
+
+    def connected(self, x, y):
+        """Test whether `x` and `y` are in the same component/set.
+
+        If neither `x` nor `y` has been previously seen, `x` is
+        inserted into the disjoint set before `y`.
```
It's weird to me that this modifies the state which can break things like this.

```
>>> import disjoint
>>> dis = disjoint.DisjointSet()
>>> dis.merge(1, 10)
True
>>> dis.merge(2, 20)
True

>>> for a in dis:
...     if dis.connected(a, 30): print("True")
... 
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
RuntimeError: OrderedDict mutated during iteration
```

pmla replied:
Nice catch. I have changed the behaviour of `__iter__` so that it only iterates over nodes which are present at the time of instantiation. A unit test has been added too.

pmla commented:
> the `in` operator seems to be linear (via `__iter__`?)

I have added a `__contains__` method. The benchmark looks better now:
```
[ 66.67%] ··· sparse_csgraph_disjoint_set.Bench.time_contains                                                               ok
[ 66.67%] ··· ======= =========
                 n             
              ------- ---------
                100    495±3ns 
                1000   497±3ns 
               10000   498±2ns 
              ======= =========
```

@sethtroisi are there any other standard class methods we should implement?

pmla commented on scipy/sparse/csgraph/_disjoint_set.py:
```diff
@@ -0,0 +1,175 @@
+
+    >>> disjoint_set.merge(1, 2)
+    True
+    >>> disjoint_set.merge(3, 'a')
```
I changed the behaviour so that `disjoint_set.merge(3, 10)` will now raise `KeyError: 10`.
Elements must now be explicity added in the `__init__` or `add` methods.

sethtroisi commented on scipy/sparse/csgraph/_disjoint_set.py:
```diff
@@ -0,0 +1,175 @@
+            x = parents[x]
+        return x
+
+    def add(self, x):
```
There don't seem to be any tests for this.

pmla replied:
Tests have been added.

pmla commented on benchmarks/benchmarks/sparse_csgraph_disjoint_set.py:
```diff
@@ -0,0 +1,36 @@
+        return [dis[i] for i in self.nodes]
+
+    def time_contains(self, n):
+        return self.nodes[0] in self.premerged
```
I have added tests along these lines.

pmla commented on benchmarks/benchmarks/sparse_csgraph_disjoint_set.py:
```diff
@@ -0,0 +1,36 @@
+        dis = self.disjoint_set
+        for a, b in self.edges:
+            dis.merge(a, b)
+
```
Great suggestion. I have added more benchmark cases.

pmla commented:
Thanks for another great review @sethtroisi.

I have implemented the `__len__` method and removed the `n_elements` attribute. I think this is the correct use of `__len__`, since it is intuitive that `len(disjoint_set) == len(list(disjoint_set))`.

The `str` method is a good suggestion. I will save it for later though.

pmla commented:
My main remaining concern is where this class belongs. It feels out of place in `sparse.csgraph` and `sparse`, since it is not a sparse matrix class. `cluster` is where I feel it most naturally belongs, but I would like to be able to import it from `spatial` too. If we put it in `_lib` it isn't exposed publicly.

Is the solution to put it in `_lib` and expose it via `cluster`? I'm just guessing here, so suggestions from experienced hands are most welcome.

sethtroisi commented on benchmarks/benchmarks/sparse_csgraph_disjoint_set.py:
```diff
@@ -19,18 +19,41 @@ def setup(self, n):
+        assert self.nodes[n // 2] in self.pre_merged
+        assert self.nodes[-1] in self.pre_merged
+
+        # Test for absense
+        assert None not in self.pre_merged
+        assert "dummy" not in self.pre_merged
+        assert (1, 2, 3) not in self.pre_merged
```
can these be moved to a separate method? I want the benchmark to verify that they take the same amount of time.

pmla replied:
Got it. I have split the test in two.

On my laptop they do not take the same amount of time:
```
[ 58.33%] ··· sparse_csgraph_disjoint_set.Bench.time_absence                                                                ok
[ 58.33%] ··· ======= ==========
                 n              
              ------- ----------
                100    699±20ns 
                1000   678±30ns 
               10000   678±20ns 
              ======= ==========

[ 66.67%] ··· sparse_csgraph_disjoint_set.Bench.time_contains                                                               ok
[ 66.67%] ··· ======= =============
                 n                 
              ------- -------------
                100    1.25±0.02μs 
                1000   1.27±0.01μs 
               10000   1.36±0.05μs 
              ======= =============
```

ev-br commented:
> cluster is where I feel it most naturally belongs, but I would like to be able to import it from spatial

So you want to be able to import the same thing from both cluster and spatial? That would be a wee bit confusing, I'd say.

pmla commented:
> So you want to be able to import the same thing from both cluster and spatial?

I would like to put this class in cluster. Then, from (internal) code in spatial, I would like to be able to write `from scipy.cluster import DisjointSet`.

ev-br commented:
Then indeed, maybe put it into _lib, like you mentioned

sethtroisi commented:
I've reviewed several times and don't have any issues (just the future TODO for adding `__str__`,` __bool__`, and `__eq__`)

Happy to see this good in soon if we could get another LGTM.

sethtroisi commented on scipy/_lib/_disjoint_set.py:
```diff
@@ -0,0 +1,174 @@
+        self.n_subsets = 0
+        self._sizes = {}
+        self._parents = {}
+        self._indices = collections.OrderedDict()
```
consider adding a comment that `_indices` tracks the order each item was added.

pmla replied:
Done.

sethtroisi commented on scipy/_lib/_disjoint_set.py:
```diff
@@ -0,0 +1,174 @@
+
+        # find by "path halving"
+        parents = self._parents
+        while self._indices[x] != self._indices[parents[x]]:
```
Is this ever true with `x != parents[x]`?
Would `hash(x) != hash(parents[x])` be faster?

pmla replied:
`x != parents[x]` fails if `x` is e.g. a numpy integer. fuglede and I discussed this at some length above.

`hash` is susceptible to collisions:
```
>>> hash((1,0,0))
2528505496374819208
>>> hash((1,-2,-2))
2528505496374819208
```
(from a [lively debate](https://bugs.python.org/issue34751))

sethtroisi replied:
Thanks for linking. I have more respect for Raymond, Tim, and Python core after reading about that journey.

Seem fudlede and you hashed this out already, I would not have thought of np.int / ndarray so great catch on `__eq__` being wonky.

`hash(x) == hash(parents[x]) and x == parents[x]` might be better but it isn't enought to overcome :/

fuglede replied:
The `hash and __eq__` approach should be equivalent to using `_indices`, cf. the discussion in the other comment: If the hashes collide, an equality comparison is forced already in `not in`.

pmla commented on scipy/sparse/csgraph/tests/test_disjoint_set.py:
```diff
@@ -15,119 +15,150 @@ def generate_random_token():
+def get_elements(n):
+    elements = set()
+    for element in generate_random_token():
+        if element not in elements:
```
Good point. I started out with a list and then changed it to a set.
I have removed it.

pmla commented on scipy/sparse/csgraph/tests/test_disjoint_set.py:
```diff
@@ -15,119 +15,150 @@ def generate_random_token():
+    return list(elements)
 
 
 @pytest.mark.parametrize("n", [10, 100])
```
Thanks. Fixed.

pmla commented:
I have moved the public interface to `scipy.cluster.hierarchy`. The class itself is in `scipy._lib`, so it will still be accessible to internal scipy code without introducing new dependencies.

Once this is green I am happy with it.

pmla commented:
At the moment I think we just need a maintainer with sufficient "ownership" of `scipy.cluster.hierarchy` to approve this.

sethtroisi commented:
@pmla I'm new so not sure how to help get this committed; I'd personally fall back to pinging rgommers when I'm blocked. 

If It would be helpful to summarize on the mailing list or something else I'm happy to do that.

fuglede commented on scipy/_lib/_disjoint_set.py:
```diff
@@ -0,0 +1,171 @@
+        self.n_subsets = 0
+        self._sizes = {}
+        self._parents = {}
+        # _indices tracks the element insertion order
```
Obvious as it may be, it would have helped me ever so slightly if it was spelled out that the `OrderedDict` is necessary only for `iter(self._indices)` below.

pmla replied:
OK - I have added this comment:
```
        # _indices tracks the element insertion order - OrderedDict is used to
        # ensure correct ordering in `__iter__` and for tie-breaking in
        # `merge`.
```

fuglede replied:
A bit nitpicky I know, but from what I can tell, the fact that it's an `OrderedDict` is not actually used during tie-breaking: There, all your using is that `self._indices[x]` counts the insertion order, which you would also achieve with an ordinary `dict`.

pmla replied:
I went with an `OrderedDict` because this behaviour has only been "official" since python 3.7. From the release notes:

> Changed in version 3.7: Dictionary order is guaranteed to be insertion order. This behavior was implementation detail of CPython from 3.6.

Recommendations on the internet suggest using `OrderedDict` for python 3.6 because other python implementations might not have this behaviour, and SciPy has not yet dropped support for 3.6.

pmla replied:
Does that make sense?

fuglede replied:
I assume you mean the part of `merge` relevant to tie-breaking defined in the conditional

```python
if (sizes[xr], self._indices[yr]) < (sizes[yr], self._indices[xr]):
```

right? If so, that would work just the same if `self._indices` were a `dict`, even prior to CPython's adoption of the ordered dict implementation; all that matters is the value of `self._indices[xr]` resp. `...[yr]`, and you're setting that explicitly.

pmla replied:
Oh, right - I'm on the same page now. You are absolutely correct. I have fixed the comment.

fuglede commented on scipy/_lib/_disjoint_set.py:
```diff
@@ -0,0 +1,171 @@
+            x = parents[x]
+        return x
+
+    def add(self, x):
```
`add` will potentially mess up the object's state if `x` (or more precisely, something equal to `x`) already exists in the disjoint set. Might that be worth an explicit check (and corresponding tests)?

pmla replied:
Great catch. This is an absolute howler of a bug.

I have fixed the function and added a test that `add` is idempotent.

fuglede commented on scipy/cluster/tests/test_disjoint_set.py:
```diff
@@ -0,0 +1,163 @@
+
+    assert elements == list(dis)
+    roots = [dis[i] for i in elements]
+    assert all([elements[0] == r for r in roots])
```
Purely a matter of taste, but `[` and `]` are redundant

pmla replied:
Fixed.

fuglede commented on scipy/cluster/tests/test_disjoint_set.py:
```diff
@@ -0,0 +1,163 @@
+
+        assert elements == list(dis)
+        roots = [dis[i] for i in elements]
+        expected_indices = np.arange(n) - np.arange(n) % (2 * k)
```
Just a sanity check: `expected_indices` will always be all zeros which means that this example is a bit trivial at the end of the day, but maybe this is intended?

pmla replied:
It is all zeros on the last iteration only. On the preceding iterations it is not though - I have just double checked this.

fuglede replied:
Ah, right, my bad, sorry about the noise.

fuglede commented on scipy/cluster/tests/test_disjoint_set.py:
```diff
@@ -0,0 +1,163 @@
+
+
+@pytest.mark.parametrize("n", [10, 100])
+def test_linear_union_sequence(n):
```
It could be nice to have a version of this test that merges in a different order than the one that leads to the lower insertion order being the root. That is, a test that explicitly shows that set size goes before insertion order when deciding the new root. For instance that looping in the reverse order would lead to `n - 2` being the root for all elements. Or something as simple as showing that

```python
dis = DisjointSet([1, 2, 3])
dis.merge(2, 3)
dis.merge(1, 2)
```

leads to `2` being the root, but

```python
dis = DisjointSet([1, 2, 3])
dis.merge(1, 2)
dis.merge(2, 3)
```

causes `1` to be the root.

pmla replied:
Done.

fuglede reviewed (commented):
If it helps, I went through everything again and only added a few somewhat minor remarks.

I wanted to check the docs as well but unfortunately it looks like the CircleCI artifact is gone now, so that was a bit too much trouble.

pmla commented on benchmarks/benchmarks/cluster_hierarchy_disjoint_set.py:
```diff
@@ -0,0 +1,59 @@
+        assert self.nodes[-1] in self.pre_merged
+
+    def time_absence(self, n):
+        # Test for absense
```
Fixed.

pmla commented on scipy/_lib/_disjoint_set.py:
```diff
@@ -0,0 +1,171 @@
+        merged : bool
+            True if `x` and `y` were in disjoint sets, False otherwise.
+        """
+        x = self[x]
```
OK - I have called them `xr` and `yr` for brevity.

pmla commented:
Thanks for the review @fuglede. The bug you found was critical.

fuglede commented on scipy/_lib/_disjoint_set.py:
```diff
@@ -0,0 +1,175 @@
+    -----
+    This class implements the disjoint set [1]_, also known as the *union-find*
+    or *merge-find* data structure. The *find* operation (implemented in
+    `__getitem__`)* implements the *path halving* variant. The *merge* method
```
Is the first asterisk supposed to be there? Renders as follows in the docs:

![image](https://user-images.githubusercontent.com/6169306/92332581-40fc2280-f07f-11ea-941b-b82115a23ca3.png)

pmla replied:
Whoops, no.
Fixed.

fuglede commented on scipy/cluster/hierarchy.py:
```diff
@@ -127,13 +134,14 @@
 _EUCLIDEAN_METHODS = ('centroid', 'median', 'ward')
 
-__all__ = ['ClusterNode', 'average', 'centroid', 'complete', 'cophenet',
+__all__ = ['ClusterNode', 'DisjointSet', 'average', 'centroid', 'complete', 'cophenet',
```
Just to save a bit of time while waiting for the next batch of CI results: [Linter thinks](https://travis-ci.org/github/scipy/scipy/jobs/724697851) this line is too long.

pmla replied:
Fixed.

pmla commented:
Depending on what is decided on [support for python 3.6 in SciPy 1.6.0](https://mail.python.org/pipermail/scipy-dev/2020-August/024339.html), we can replace the `OrderedDict` with a standard `dict`.

grapheo12 commented:
Hello, newbie here!
I had a question: Why don't we leverage multiprocessing while finding the parent of a node, in `connected` and `merge` methods? While computing parents of two very large, uncompressed, disjoint sets, we can benefit from multiprocessing.
We can get around the chance of race conditions by using appropriate node based locks.

There seems to be a very recent work related to this subject: https://arxiv.org/abs/2003.01203

rgommers commented:
> API-wise, if I read @rgommers' comment on the mailing list correctly, he considered having it live in sparse directly, as opposed to sparse.csgraph; this might simplify the dependency graph.

I meant `sparse.csgraph`, that seemed to be the closest fit. It's very similar to code already present there; reading the Wikipedia page that's the reference in the docstring, that mentions minimum spanning tree of a graph and connected components (both of which we have in `csgraph`); finally there's discussion here about the equivalent being in NetworkX and the Wikipedia page mentions another equivalent in the Boost Graph Library.

So I'm slightly surprised by the choice for `cluster.hierarchy`. That said, I haven't reviewed this PR in detail - if reviewers agree that `cluster` is a good choice, I'm happy with that as well. 

> At the moment I think we just need a maintainer with sufficient "ownership" of `scipy.cluster.hierarchy` to approve this.

`cluster` doesn't really have a dedicated maintainer unfortunately. It's mostly been @nmayorov and me who have reviewed PRs for it.

rgommers commented:
@pmla you may want to rewrite the commit history. Or we can squash-merge.

charris commented:
Cluster makes sense to me, clustering is what I've always used the algorithm for, it provides the equivalence classes of an equivalence relationship,

pmla commented:
@rgommers @charris Thanks for weighing in. Cluster it is. I will leave it in `cluster.hierarchy` since `cluster.vq` is not relevant here.

@sethtroisi would you like to merge this? A squash-merge has been recommended by Ralf.

sethtroisi commented:
@pmla Happy to merge this! Thanks for the work to make this better along the way.

rgommers commented:
Very nice! Could you add a release note for this in https://github.com/scipy/scipy/wiki/Release-note-entries-for-SciPy-1.6.0 @pmla?

pmla commented:
@rgommers Yep. Done.
